server/decorators.go

```diff
@@ -38,3 +43,28 @@ func Delay(delay time.Duration, h http.HandlerFunc) http.HandlerFunc {
 		time.Sleep(delay)
 	}
 }
+
+func Cache(h http.HandlerFunc) http.HandlerFunc {
+	messages := make(map[string]greet.Greet)
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		// Read body into buffer to avoid consuming it - https://stackoverflow.com/questions/23070876/reading-body-of-http-request-without-modifying-request-state
+		bufferBody, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			return
+		}
+
+		r.Body = ioutil.NopCloser(bytes.NewBuffer(bufferBody))
+		defer h.ServeHTTP(w, r)
```

Comment:
Os ha pasado a muchos esto: si no hacéis h.Server(w, r) no os hace faltas volver a escribir el Body.

Reply:
El ServeHTTP es cuando vais a pasar otro handler (h) y el que lo envuelve necesita llamarlo para que continúe la cadena